Add a cron service and command with cron expressions

fritzmg opened the pull request:
This is a revamp of my original PR https://github.com/contao/contao/pull/708. The basic features and service tag options are the same - with the addition of being able to use full CRON expressions, instead of just `minutely`, `hourly`, etc.!


### Command

Cron jobs can now be executed via the following command:

```
vendor/bin/contao-console contao:cron
```

This can be used in a minutely crontab for example. When using the command, only cron jobs with no scope definition (`null`) or the `cli` scope will be executed, jobs with the scope `web` will be skipped.


### Service Tag

The new service tag `contao.cronjob` has the following options:

| Option | Description |
| --- | --- |
| `interval` | Can be `minutely`, `hourly`, `daily`, `weekly`, `monthly`, `yearly` or a full CRON expression, like `*/5 * * * *`. |
| `method` | Will default to `__invoke` or `onMinutely` etc. when a named interval is used. Otherwise a method name has to be defined. |

_Example:_

```yml
services:
  App\Cron\ExampleCron:
    tags:
      -
        name: contao.cronjob
        interval: * 0 * * *
        method: onDaily
```


### Annotation

Tagging services is also possible via Annotations:

```php
// src/Cron/ExampleCronJob.php
namespace App\Cron;

use Contao\CoreBundle\ServiceAnnotation\CronJob;
use Terminal42\ServiceAnnotationBundle\ServiceAnnotationInterface;

/**
 * @CronJob("* 0 * * *")
 */
class ExampleCronJob implements ServiceAnnotationInterface
{
    public function __invoke(): void
    {
        // …
    }
}
```

When the annotation is used on the class, either the `__invoke` method will be used - or an auto generated method name (e.g. `onMinutely`), if present. Basically the same concept as in https://github.com/contao/contao/pull/1063.

_Note:_ If you need an interval like `*/5 * * *` you need to escape either the `*` or `/` with `\`, since `*/` would close the PHP comment. This is what line 56 in the `Cron` annotation class is for. @aschempp may be this is something, that the ServiceAnnotationBundle should handle by itself?


### Front End Cron & Route

The `/_contao/cron` route still exists but now uses the new `Cron` service instead of the `Contao\FrontendCron` controller class. Same with the `Contao\CoreBundle\EventListener\CommandSchedulerListener` (poor man's cron). Cron jobs tagged with the scope `cli` will __not__ be executed for either.


### Additional Dependency

I am using [dragonmantank/cron-expression](https://github.com/dragonmantank/cron-expression) to parse CRON expressions and check if a cron job is, _or was_, due. I did not use [cron/cron](https://github.com/cron/cron) because it lacks the latter functionality. The Cron within Contao might or will be executed in irregular intervals, thus it needs to be checked whether a cron job with a defined interval would currently be _overdue_ relative to its last execution.

_Example:_ say you have cron job with an interval of `*/5 * * * *` and you are still using the front end command scheduler. This cron job would be due every 5 minutes of every hour - e.g. `00:00`, `00:05`, `00:10` etc. However, there might not be a request which would execute the Cron service every minute. If the Cron service is executed at `00:03:58` for example and the cron job has not been executed before, the cron job will be executed right away. Then if the next run of the Cron service is at `00:07:10` the cron job will be executed again, since its next run would have been at `00:05:00`, relative to its last execution time. If another request is made at `00:09:46`, the cron job will not be executed, since its next scheduled execution time would be at `00:10:00` - etc.

As for the package itself: it looks well maintained, has lots of installs and dependents - and apparently Laravel uses it too.


### New Entity

The PR introduces a new `Cron` entity, which simply saves the "name" of a cron job (consisting of the class name and method) and its last execution.

![tl_cron](https://user-images.githubusercontent.com/4970961/70894253-f4eae880-1fec-11ea-8d59-23b2068127ee.png)


### Backwards Compatibility

I have completely deleted the `tl_cron` DCA and am now using the `tl_cron` table for the new format (via an entity). This is probably not ideal for BC? ;)

If we want to keep the old `tl_cron` table as is, we could use a new `tl_cron_runs` table instead. In that case: should we still update the values in the `tl_cron` table? May be simply through the legacy cron.

### Scope

In some cases a cron job might want to know in which "scope" it is executed in - i.e. as part of a front end request as part of the cron command on the command line interface. The `Cron` service will pass a scope parameter to the cron job's method.

```php
namespace App\Cron;

use Contao\CoreBundle\Cron\Cron;

class HourlyCron
{
    public function __invoke(string $scope): void
    {
        // Do not execute this cron job in the web scope
        if (Cron::SCOPE_WEB === $scope) {
            return;
        }

        // …
    }
}
```

fritzmg committed:
add cron command and scheduler using CRON expressions

fritzmg committed:
remove superfluous use statement

Toflar commented on core-bundle/src/Entity/Cron.php:
```diff
@@ -0,0 +1,74 @@
+     */
+    protected $lastRun;
+
+    public function __construct(string $name, \DateTime $lastRun = null)
```
We should probably typehint for `DateTimeInterface` instead. A `DateTimeImmutable` object would be fine here too, I suppose :)

fritzmg replied:
Sure, I have no preference for either - which one should I use?

Toflar commented on core-bundle/src/Repository/CronRepository.php:
```diff
@@ -0,0 +1,59 @@
+    {
+        $table = $this->getClassMetadata()->getTableName();
+
+        $this->connection->exec("LOCK TABLES $table WRITE, $table AS t0 WRITE, $table AS t0_ WRITE");
```
Why do you have to lock aliases here?

fritzmg replied:
The `RememberMeRepository` does the same, otherwise there will be an error. 

```
An exception occurred while executing 'SELECT t0.id AS id_1, t0.name AS name_2, t0.lastRun AS lastRun_3 FROM tl_cron t0 WHERE t0.name = ? LIMIT 1' with params ["App\\Cron\\ExampleCliCron::__invoke"]:

SQLSTATE[HY000]: General error: 1100 Table 't0' was not locked with LOCK TABLES 
```

Weirdly, I had to use `t0` instead of `t0_` 🤷‍♂ .

Toflar commented on core-bundle/src/Repository/CronRepository.php:
```diff
@@ -0,0 +1,59 @@
+        $this->connection->exec('UNLOCK TABLES');
+    }
+
+    public function persist(CronEntity ...$entities): void
```
I don't like this method name. Imho it should be `persistAndFlush()` or something similar. Everybody working with Doctrine knows exactly that `persist()` does not persist to the database automatically but your method does. That's irritating to me :)

fritzmg replied:
Well, it is called `persist` in the already existing `RememberMeRepository`, so ... ;)

Toflar replied:
That's not good either then.

aschempp replied:
Actually I don't think the repository should persist or flush at all 🙈

Toflar commented on core-bundle/src/Resources/contao/controllers/FrontendCron.php:
```diff
@@ -26,82 +31,13 @@ class FrontendCron extends Frontend
-
-		while ($objLock->next())
+		// Do not run if there is POST data
+		if (empty($_POST))
```
To be consistent with the `_cron` route and we're now importing the container anyway, can we use the current request and check if it's a `GET` request maybe?

Toflar replied:
We could maybe even forward to the `/contao/_cron` route but that might be a bit too much just for BC.

leofeyer replied:
This check is weaker than the check in the `FrontendController` class, where we explicitly require the method to be `GET`:

```php
$request->isMethod(Request::METHOD_GET)
```

I think we should do the same here – unless we decide to forward as suggested by @Toflar.

aschempp replied:
You're aware this method is not used anymore at all?

fritzmg replied:
Yeah, this class is not used at all anymore, that's why I did not put any work into it.

leofeyer replied:
Agreed. 👍

Toflar reviewed (commented):
Looks really good implementation-wise! I like the concept. I added a few remarks and here are some more general ones:

* I don't thing the `ServiceAnnotationBundle` should handle comment escaping. If any, then the annotation extrator itself. But I don't think that's really possible. I think escaping is just fine and it's a matter of adding it to the documentation.

* I'm fine with the dependency. I checked it myself and there's nothing to worry about.

* I'm also fine dropping the existing `tl_cron` table and replacing it with an entity. I don't see any reason why anyone would've ever extended that table or relied on it. And if so, that sounds like an architectural mistake to me that can be easily fixed by providing own ways to store meta data.

* I think this PR should wait for @ausi's migration framework PR to be merged into master so that the migration is already correct here.

So to sum up: Excellent work @fritzmg - I really like this one as well!

fritzmg commented:
> * I think this PR should wait for @ausi's migration framework PR to be merged into master so that the migration is already correct here.

Agreed, I was planning on doing that anyway, forgot to mention it. I added the migration step as a "reminder" of sorts.

fritzmg committed:
implement changes reported by code analyzer

fritzmg commented:
The code analyzer reports the following:

```
 ------ -------------------------------------------------------------------- 
  Line   tests/DependencyInjection/Compiler/AddCronJobsPassTest.php          
 ------ -------------------------------------------------------------------- 
  31     Call to an undefined method                                         
         Symfony\Component\DependencyInjection\ContainerBuilder::method().   
  37     Call to an undefined method                                         
         Symfony\Component\DependencyInjection\ContainerBuilder::expects().  
  242    PHPDoc tag @return contains unresolvable type.                      
 ------ -------------------------------------------------------------------- 
```
Not sure what do do here - am I using outdated phpunit methods?

aschempp commented on core-bundle/src/Cron/Cron.php:
```diff
@@ -0,0 +1,139 @@
+class Cron
+{
+    /**
+     * @var string
```
Nitpicking: I don't think constants ever have docblocks

fritzmg replied:
But constants can be strings, floats, integers or arrays - shouldn't this be defined this way?

aschempp commented on core-bundle/src/Cron/Cron.php:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * @var array
+     */
+    private $crons = [];
```
to be consistent, shouldn't this be named `$cronJobs`?

aschempp commented on core-bundle/src/Cron/Cron.php:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Add a cron service.
+     *
+     * @param string $interval The interval as a CRON expression
```
why is this `@param` needed?

fritzmg replied:
To provide additional info about the parameter (i.e. that it is supposed to be a "CRON expression").

aschempp commented on core-bundle/src/Cron/Cron.php:
```diff
@@ -0,0 +1,139 @@
+     *
+     * @param string $interval The interval as a CRON expression
+     */
+    public function addCronJob($service, string $method, string $interval, int $priority = 0, string $scope = null): void
```
is `$service` a `callable` or `object` (as of PHP 7.2 if I'm right)? Maybe we pass use a `callable` instead (which in most cases will be just the service object for `__invoke`, or an array of object and method)?

fritzmg replied:
Hmm... well I need the method name for the `name` in the table anyway. Or should we ommit `::method` for the name, if the object is a `callable` (and thus only save the class name as the name of the cron job)?

aschempp reviewed (commented):
Looking very good indeed! 🎉 

Some thoughts:
 - I think the `@Cron` annotation should be named `@CronJob` to be consistent.
 - I would also name the table `tl_cron_job`. There's no value in keeping the same table name, but it can be confusing. To be honest, I did sometimes use that table in my extensions, and using a new name would at least make it obvious that this really has changed.
 - I would suggest to not have `onMinutely` etc. auto-detection for method names. The implementation in the mentioned PR was only for backwards compatibility. A class annotation should always be for the `__invoke` method if no method name is given.
 - One conflict I can foresee: what if I want to run the same class & method multiple times, by defining multiple cron job times? This would lead to conflicts about lastRun in the database table, wouldn't it?

fritzmg commented:
> * One conflict I can foresee: what if I want to run the same class & method multiple times, by defining multiple cron job times? This would lead to conflicts about lastRun in the database table, wouldn't it?

Very good point, I haven't thought of that, since I always only considered the annotation (where this wouldn't be possible?). But it is perfectly possible to use multiple tags for the same service (and thus the same method) of course.

To fix this, I would also save the defined interval in the `tl_cron`/`tl_cron_job` table (and create a unique index over `name,interval`).

aschempp reviewed (commented):
More thoughts: I think the current `Cron` implementation could be optimized:

1. Why is the priority not applied in the compiler pass and therefore *automatically* correct in the `addCronJob` method? Do we really need another priority *inside* the cron handler?
2. Maybe we should use a data object for the cron definition instead of an array?
3. if one of the cronjob currently fails, all jobs will be marked as executed because that's done before running them. This certainly will be a problem and should be re-thought.

I can't remember why we had the cron scopes at all? I mean how can I ever _want_ a cron job not to run when I run my cronjobs? It would basically mean I will always need to setup the CLI **and** web callbacks to execute everything? I know there might be jobs that can't run on the CLI for legacy reasons, but shouldn't that be deprecated/disallowed in the new setup?

fritzmg commented:
> 1. Why is the priority not applied in the compiler pass and therefore _automatically_ correct in the `addCronJob` method? Do we really need another priority _inside_ the cron handler?

True, this could already be sorted in the in the compiler pass.


> 2. Maybe we should use a data object for the cron definition instead of an array?

I am all for that. Do we use anything like that in the core yet, so that I have a reference?

> 3. if one of the cronjob currently fails, all jobs will be marked as executed because that's done before running them. This certainly will be a problem and should be re-thought.

I would simply use a try/catch block within the cron job execution loop, creating a log entry when a job fails. This way the other jobs can still be executed and all of them being marked as executed will be correct then.

There might be edge cases where the whole cron service fails and thus jobs are being marked as run even though they weren't. I'd rather not introduce an additional database flag or anything like that showing the current state of the job, due to locking and concurrency.

aschempp commented:
> > 1. Maybe we should use a data object for the cron definition instead of an array?
> 
> I am all for that. Do we use anything like that in the core yet, so that I have a reference?

The migrations command PR does: https://github.com/contao/contao/pull/709/files#diff-68d11ab2b9b2006b39d4cb82a8771588

> I would simply use a try/catch block within the cron job execution loop, creating a log entry when a job fails. This way the other jobs can still be executed and all of them being marked as executed will be correct then.

try-catch and logging would absolutely make sense to me. Be aware that we might want to log to `tl_log` (`ContaoContext`).

> There might be edge cases where the whole cron service fails and thus jobs are being marked as run even though they weren't. I'd rather not introduce an additional database flag or anything like that showing the current state of the job, due to locking and concurrency.

How about we only update the cronjob table after the job has been executed? Or you mean you don't want to lock the tables for that long period?

> To fix this, I would also save the defined interval in the tl_cron/tl_cron_job table (and create a unique index over name,interval).

Is there any chance we can use this to actually manipulate the cron job run times as an application maintainer (which is a feature I'd like to see, but has low prio)? 🙃

fritzmg commented:
> How about we only update the cronjob table after the job has been executed? Or you mean you don't want to lock the tables for that long period?

Yeah, I did not want to lock the tables for the complete duration of all cron jobs. Suppose you set `contao:cron` to execute minutely in your crontab... what happens if one of those cron jobs runs for an hour - which would mean the table is locked for an hour?

fritzmg commented:
> Is there any chance we can use this to actually manipulate the cron job run times as an application maintainer (which is a feature I'd like to see, but has low prio)? 🙃

Hm, couldn't you just implement your own compiler pass (which is run before `AddCronJobsPass`) and change the service tag accordingly?

aschempp commented:
> Yeah, I did not want to lock the tables for the complete duration of all cron jobs. Suppose you set `contao:cron` to execute minutely in your crontab... what happens if one of those cron jobs runs for an hour - which would mean the table is locked for an hour?

Hmm, not sure if that solves it actually. The minutely cronjob would be started a minute later, even though it is still running for an hour?

> > Is there any chance we can use this to actually manipulate the cron job run times as an application maintainer (which is a feature I'd like to see, but has low prio)? 🙃
> 
> Hm, couldn't you just implement your own compiler pass (which is run before `AddCronJobsPass`) and change the service tag accordingly?

Very true, but complicated 😂 
I liked the idea of cron/cron where a app user can decide what job to run when. In the Managed Edition®, it mostly makes sense to have it defined by the dev, but maybe I wanna change it. But nevermind, thats not really in this PRs scope.

fritzmg commented:
> Hmm, not sure if that solves it actually. The minutely cronjob would be started a minute later, even though it is still running for an hour?

Example: you have set up to run the Contao cron every minute. There is a job that runs for 60 minutes. Every minute, this job is executed and then runs in parallel with all the other executions, if the database table is not locked during the whole execution of all crons. This would be the current behaviour and also the old behaviour.

However, if you _do_ lock the table for the whole execution, then the Contao cron process is still spawned every minute. However, the second run has to wait, until the first one is finished, then a minute later, a new process is spawned that also has to wait until the table is not locked anymore, etc. This would spawn infinite processes. Unless the process fails, when a database lock cannot be achieved? I don't actually know what happens then. I'd have to test.

Anyhow: we could also save the state of each cronjob in the database and thus automatically skip its execution, if it is still running. So `tl_cron_job.state` would either have nothing, `running`, `finished` or `failed` in it and a job (with a given interval) can only be executed again, if its state is either empty or not `running`.


> Very true, but complicated 😂
> I liked the idea of cron/cron where a app user can decide what job to run when. In the Managed Edition®, it mostly makes sense to have it defined by the dev, but maybe I wanna change it. But nevermind, thats not really in this PRs scope.

Hm, what about something like this:

```php
class CronInterface
{
    /**
     * Returns all added cron jobs.
     */
    public function getCronJobs(): array;

    /**
     * Returns a specific cron job, given its name and interval.
     * If the interval is ommitted, the first cron job found with the given name will be returned.
     */
    public function getCronJob($name, $interval = null): ?CronJobInterface
}
```

```php
class CronJobInterface
{
    /**
     * Returns a callable or array.
     */
    public function getService()

    public function getName(): string
    public function getInterval(): string
 
    /**
     * Adjust the interval of the cron job.
     */
    public function setInterval(string $interval): self
}
```

Haven't really looked at how `cron/cron` handles this though :)

aschempp commented:
I just discussed my thoughts and issues with @fritzmg, here are some details:

 1. We discussed the issue of jobs running concurrently, meaning if a minutely job runs longer than one minute, it could be running twice. However, we both agreed this is not really an issue, or not one that we (Contao) should fix. The exact same would apply to a Crontab, it would simply run the process twice and not care about already running jobs.

 2. I complained about the scope idea, because I think 99% of devs should not think about it. The (only) viable solution mentioned by @Toflar (listening in on my call) was a case of a job that knows it's going to run for longer than the web process can handle, e.g. to behave differently in that case. We agreed that it would be better to check `PHP_SAPI` or have another *runtime information* to know if the command or the web job is running. It's not something one should differentiate in the annotation.

 3. There are two concerns remaining: *custom scheduling for jobs* and *execution time*

    - **Custom scheduling for jobs:** it would be nice if it was possible to override the annotation when a job should be executed. E.g. if Contao does logrotation once a week, it might make sense to run that job once a day instead.

        Solving this is actually pretty simple: as this is an advanced feature, I'd assume one would have a Crontab available. So instead of running the `contao:cron` command, one should be able to run `contao:cron --job=MySuperJob::method`. This way I can execute each job at the time I want, and possibly even in parallel by scheduling all of them individually in Crontab.

    - **Execution time:** Currently, we're fetching all current jobs from the DB, mark them as *executed now* and then run them sequentially. Apart from issues if one job quits the app, this also means a second job might run 5 minutes later, if the first job takes 5 minutes to complete.

        Given the above `contao:cron --job=…` option, the most simple solution would be to actually start a background process for every cron job, when `contao:cron` is executed without arguments. This means one process can die without affecting the others, and they are actually run in parallel and asynchronous. That's obviously something that only works on the command line and not in the web process.


As discussed with @fritzmg, the two things in point 3 are not necessarily part of this PR and can be added/improved in a later implementation (possibly for Contao 4.10). We *just* need to make sure the annotations and interfaces are correct in the first place, meaning point 2 needs to be addressed before merging this.

fritzmg committed:
remove priority and scope

fritzmg committed:
add simple test for CronCommand

fritzmg committed:
rename CronJobs

fritzmg commented:
> * I would suggest to not have `onMinutely` etc. auto-detection for method names. The implementation in the mentioned PR was only for backwards compatibility. A class annotation should always be for the `__invoke` method if no method name is given.

But this isn't just about service annotations. If you define your service tag in your service configuration, it would still be useful to not need to define the method name as well? Or should this also be restricted to only the `__invoke` method?

fritzmg committed:
implement various improvements

fritzmg committed:
fix merge error

fritzmg committed:
remove definition tests

fritzmg commented:
* I have removed the scope parameter in e7fc60e4aa663ddd2fe6df7596b34aebb7f175a8 for now.
* Since we want to execute the jobs in parallel in the future, I have also removed the priority parameter, since that wouldn't make any sense then anymore.
* `tl_cron` is now simply dropped by default (as the DCA is missing) and the entity uses `tl_cron_job` instead.
* All the relevant classes are now called `CronJob` instead of `Cron` (ce61eb03010734eca148f48719118a6feed92de2).
* `DateTimeInterface` instead of `DateTime` is now used (f939b08ec7f3ac3be321dec58c5cbbfb5325fd17).

fritzmg committed:
fix unit tests

fritzmg committed:
update database dump for functional tests

leofeyer commented:
Contao 4.9 is now in the review phase in which we try to review and merge the remaining pull requests that have been created during the development phase. Your pull request does not yet meet the requirements to be reviewed though:

* Make sure that the PR is feature complete.
* Rebase the PR onto the latest master branch.
* Address the requested changes if any.
* Make sure that all CI checks are successful.

Please fix this soon if you can, because the review phase ends on January 15th and we need some time to review the PR before we can merge it.

fritzmg committed:
code style fix

fritzmg committed:
use Cron::class for reference check

fritzmg committed:
change database table for canRunDbQuery

fritzmg commented:
Regarding scope matching: initially I thought, the simplest solution for anyone who wants to implement a cron, that is only executed (or not executed) in the "web scope" is to do the following:

```php
// Do not execute cron in web scope
if (null !== $this->requestStack->getCurrentRequest()) {
    return;
}
```

Unfortunately, that will not work, since the `CommandSchedulerListener` is executed in the `kernel.terminate` event and there, the current request will not be present anymore.

fritzmg committed:
try to fix code analysis errors

fritzmg committed:
fix return type declaration in test

fritzmg committed:
add CronJob class and scope interface

fritzmg committed:
rename service annotation class to CronJob

fritzmg commented:
I have added some changes and a new proposal for scoped cron jobs in 0a619d400f7c0e38680f673fbce5aea79bad5fca:

* A new `CronJob` class now holds the service object, method name, interval and cron job name.
* The new class also handles execution of the cron job.
* A new `ScopedCronJobInterface` can now be optionally implemented, using a `setScope` method.
* The cron service will set the scope for all cron jobs and the `CronJob` class will pass this down to the service, if it is an instance of `ScopedCronJobInterface`.

@leofeyer @aschempp @Toflar wdyt?

I also renamed the annotation from `Cron` to `CronJob` and the service tag from `contao.cron` to `contao.cronjob` in e248bc7e7ed05315424531de5aa0ee7fc10fb2a9 to be consistent with the rest.

leofeyer commented:
Can you explain why we need a `LegacyCron` class which only supports a handful of hardcoded cron jobs? Can we not refactor theses methods?

fritzmg commented:
The `LegacyCron` class executes all cronjobs registered via `$GLOBALS['TL_CRON']`. It does not execute any hardcoded cron jobs. I think you confused the check for the core cronjobs: this is just there to issue a deprecation warning, whenever someone uses `$GLOBALS['TL_CRON']` instead of a tagged service. The check ensures that this deprecation warning is not issued for the legacy cronjobs of the core.

leofeyer commented:
But why? Shouldn't the core stop using `$GLOBALS['TL_CRON']` as well?

fritzmg commented:
> But why? Shouldn't the core stop using `$GLOBALS['TL_CRON']` as well?

🤷‍♂ not sure, don't we have to keep the core cronjobs as they are for BC? But the `LegacyCron` class would still be needed.

leofeyer commented:
Yes, the `LegacyCron` class needs to stay. But there should not be an exception for the core cronjobs IMHO. What do the others think?

Toflar commented on core-bundle/src/Cron/ScopedCronJobInterface.php:
```diff
@@ -0,0 +1,21 @@
+
+namespace Contao\CoreBundle\Cron;
+
+interface ScopedCronJobInterface
```
This should be named `ScopeAwareCronJobInterface` to be consistent with lots of Symfony libraries. Everything that accepts a setter is "aware" of it :)

Toflar commented on core-bundle/src/Controller/FrontendController.php:
```diff
@@ -44,13 +45,13 @@ public function indexAction(): Response
+        }
 
-        return $controller->run();
+        return new Response('', 204);
```
This should use the `Response::HTTP_NO_CONTENT` constant to be consistent.

Toflar commented on core-bundle/src/Resources/contao/controllers/FrontendCron.php:
```diff
@@ -26,82 +31,13 @@ class FrontendCron extends Frontend
-		}
-
-		return $objResponse;
+		return new Response('', 204);
```
Use the constant here too please

Toflar reviewed (commented):
Minor nitpicks. This PR is in good shape and we should push to merge this asap.
Regarding the deprecation message. I would leave it as is for now but maybe @fritzmg can work on a PR that replaces our own `TL_CRON` entries with the new framework for 4.10. Imho we can add a `// TODO: migrate our own cronjobs to new framework` comment to those exceptions and it's all good.
Let's aim for a timely solution here.

Toflar reviewed (changes requested):
Forgot to request changes, sorry.

fritzmg commented:
@Toflar thx, I will work on the changes today.

leofeyer commented on core-bundle/src/Cron/Cron.php:
```diff
@@ -0,0 +1,119 @@
+
+        /** @var array<CronJob> */
+        $cronJobsToBeRun = [];
+        $now = new \DateTime();
```
```suggestion
        $now = new \DateTimeImmutable();
```

leofeyer commented on core-bundle/src/Cron/CronJob.php:
```diff
@@ -0,0 +1,86 @@
+
+namespace Contao\CoreBundle\Cron;
+
+final class CronJob
```
We have not yet agreed on the concept of final classes, therefore the `final` keyword has to be removed (unless there are technical reasons that require it).

fritzmg replied:
But it is used here too: https://github.com/contao/contao/blob/133696aa02ef2dfb4005937e25b8c8427de885ec/core-bundle/src/Migration/MigrationResult.php#L15

Should it get removed from both?

leofeyer replied:
Yes, until we have agreed on a concept, it needs to be removed there as well. Right now, we are only using final classes for

* Anything that `implements ServiceTagInterface`
* Any `*Events` class
* The `ContaoModuleBundle` class
* Anything that `extends AbstractFragmentAnnotation`

leofeyer commented on core-bundle/src/Cron/LegacyCron.php:
```diff
@@ -0,0 +1,104 @@
+
+class LegacyCron implements ServiceAnnotationInterface
+{
+    private const ALLOWED_CRONS = [
```
As discussed, we do not need exceptions for the core cron jobs.

fritzmg replied:
Have we decided this yet? 😁 

If we only implement the core cronjobs as services in Contao `4.10`, then the LTS version will always trigger deprecation warnings, since the core still uses `TL_CRON` there.

leofeyer replied:
Yes, this will be the case. Just like it will be the case for every third-party extension that defines cron jobs. 🤷‍♂

leofeyer commented on core-bundle/src/Repository/CronJobRepository.php:
```diff
@@ -0,0 +1,61 @@
+    {
+        parent::__construct($registry, CronJob::class);
+
+        if (($connection = $registry->getConnection()) instanceof Connection) {
```
You are checking for a particular instance here but you never handle the case that the check is not true. If `$connection` is not an instance of `Connection`, the other methods of the class will fail.

fritzmg replied:
To be honest, this class is more or less a copy & paste of the existing `RememberMeRepository`. The same happens there too:

https://github.com/contao/contao/blob/133696aa02ef2dfb4005937e25b8c8427de885ec/core-bundle/src/Repository/RememberMeRepository.php#L30-L37

I don't really know what the purpose of this check is. If the `ManagerRegistry` has no valid database connection, what difference does this make for the repository? Either way, `$this->connection` will be `null`.

leofeyer replied:
This needs to be fixed in the `RememberMeRepository` as well. You can leave the class as is and add a ticket as reminder.

leofeyer commented on core-bundle/src/ServiceAnnotation/CronJob.php:
```diff
@@ -0,0 +1,48 @@
+ *     @Attribute("value", type="string", required=true)
+ * })
+ */
+final class CronJob implements ServiceTagInterface
```
`final` is correct here (just in case you wonder) 😊

leofeyer commented on core-bundle/tests/Cron/CronTest.php:
```diff
@@ -0,0 +1,174 @@
+    {
+        $repository = $this->createMock(CronJobRepository::class);
+
+        $cronjob = $this->getMockBuilder(\stdClass::class)->addMethods(['onHourly'])->getMock();
```
Why are you mocking a `stdClass` object here? Which object is run in the real application later?

fritzmg replied:
An arbitrary object. Bug I guess there is probably a better way? We can also use the test cronjob classes from the fixtures.

leofeyer replied:
What about `$this->createMock(ScopedCronJobInterface::class)`?

fritzmg replied:
That will not be enough, as I need the cron job methods as well. I cannot add additional methods of a mock object of an existing class, can I? Only via the mock builder, afaik.

leofeyer replied:
Then let's use the fixtures class.

fritzmg replied:
Another problem I am running into here: when I use

```php
$cronjob = $this->createMock(TestCronJob::class);
```

I cannot test the logging completely, as the mocked class' name will be something random (e.g. `Mock_TestCronJob_a0f1e286`) - while when using

```php
 $cronjob = $this
    ->getMockBuilder(\stdClass::class)
    ->setMockClassName('TestCronJob')
    ->addMethods(['onMinutely', 'onHourly'])
    ->getMock()
;
```

it will be exactly what I set it to (i.e. `TestCronJob`) and thus can test on what exactly the logger will log.

fritzmg replied:
Nevertheless, I adjusted the tests.

leofeyer replied:
Why do you mock the fixtures class at all? It already is a fixtures class, so you can use it directly.

fritzmg replied:
Right 😁

leofeyer commented on core-bundle/tests/Fixtures/src/Cron/TestInvokableScopedCronJob.php:
```diff
@@ -0,0 +1,26 @@
+
+use Contao\CoreBundle\Cron\ScopedCronJobInterface;
+
+class TestInvokableScopedCronJob implements ScopedCronJobInterface
```
Should be `TestInvokableScopeAwareCronJob`.

leofeyer committed:
Fix the coding style

fritzmg committed:
make CronJob method optional

fritzmg committed:
remove final from CronJob class

fritzmg committed:
rename interface to ScopeAwareCronJob

fritzmg committed:
use DateTimeImmutable

fritzmg committed:
trigger TL_CRON deprecation warning also for core cron jobs

fritzmg committed:
ammend interface renaming

fritzmg committed:
update unit tests

fritzmg committed:
remove connection checkt

fritzmg committed:
use constant for empty response

fritzmg committed:
add TODO comment

fritzmg committed:
add test for invalid scope argument

fritzmg committed:
CS fix

fritzmg committed:
change constants values

fritzmg commented:
In addition to the requested changes, I also made the `method` of a cron job completely optional now. The `CronJob` container class will now throw an `InvalidArgumentException` if no `method` is passed and the given service is not invokable.

I have also changed the values of the scope constants as I think it is bad practise for them to be the same as the constant's name itself.

aschempp commented on core-bundle/src/Cron/Cron.php:
```diff
@@ -0,0 +1,117 @@
+    /**
+     * Run all the registered Contao cron jobs.
+     */
+    public function run(string $scope = null): void
```
can the scope really be `null`?

fritzmg replied:
I suppose we can make it mandatory.

aschempp commented on core-bundle/src/Cron/Cron.php:
```diff
@@ -0,0 +1,117 @@
+
+    public function addCronJob(object $service, string $interval, string $method = null): void
+    {
+        $this->cronJobs[] = new CronJob($service, $interval, $method);
```
I think the `CronJob` object should be passed here

fritzmg replied:
Then I need some help with implementing that in the `AddCronJobsPass` - as I only have a `Reference` object to the service there.

aschempp commented on core-bundle/src/Cron/Cron.php:
```diff
@@ -0,0 +1,117 @@
+                if (null === $lastRunDate || $now >= $expression->getNextRunDate($lastRunDate)) {
+                    // Update the cron entry
+                    $lastRunEntity->setLastRun($now);
+                    $this->repository->persistAndFlush($lastRunEntity);
```
flushing in a loop does not make sense. Also, you do not need to persist entities that were found in the table, they are already in the entity manager. It should be something like this:

```php
foreach ( … ) {
    if (null === $lastRunEntity) {
        $lastRunEntity = new CronJobEntity($name);
        $this->repository->persist($lastRunEntity);
    }
}

$this->repository->flush();
```

Though I'm really not sure the entity should offer these methods, or rather inject the entity manager or get the manager from the repository like `$this->repository->getManager()->persist/flush()`

fritzmg replied:
I think I'd implement the latter, as I'd rather not want to inject the em directly.

fritzmg replied:
I decided to inject the entity manager instead, as it also feels wrong to return the entity manager from the repository (since there is already a protected getter method for that).

aschempp commented on core-bundle/src/Cron/ScopeAwareCronJobInterface.php:
```diff
@@ -0,0 +1,21 @@
+    /**
+     * Sets the scope for the cron job.
+     */
+    public function setScope(string $scope);
```
This will result in state on the object (possibly a container service).

aschempp commented on core-bundle/src/DependencyInjection/Compiler/AddCronJobsPass.php:
```diff
@@ -0,0 +1,116 @@
+            return;
+        }
+
+        $serviceIds = $container->findTaggedServiceIds('contao.cronjob');
```
Should we support priority here? So two jobs running at the same time can theoretically have a priority? I think it can't harm…

fritzmg replied:
I removed the priority again as we want to execute the jobs in parallel in the future.

Toflar replied:
Priorities make no sense in cronjobs, native ones don't have one either.

aschempp commented on core-bundle/src/DependencyInjection/Compiler/AddCronJobsPass.php:
```diff
@@ -0,0 +1,116 @@
+                    throw new InvalidDefinitionException(sprintf('The contao.cronjob definition for service "%s" has an invalid interval expression "%s"', $serviceId, $interval));
+                }
+
+                $definition->addMethodCall('addCronJob', [new Reference($serviceId), $interval, $method]);
```
I think you should create the `CronJob` instances here. To create them lazily, this should be possible using the container.

```php
$definition->addMethodCall(
    'addCronJob', 
    [
        new Definition(CronJob::class, [new Reference($serviceId), $interval, $method]
    ]
);
```

fritzmg replied:
Ah thank you 😁

aschempp commented on core-bundle/src/Entity/CronJob.php:
```diff
@@ -0,0 +1,74 @@
+    public function __construct(string $name, \DateTimeInterface $lastRun = null)
+    {
+        $this->name = $name;
+        $this->lastRun = $lastRun ?? new \DateTime();
```
is this correct? The job was not run when the object is initialized?

fritzmg replied:
If you pass `null`, then the current date and time will be used. So the job is always considered run, when the entity is initialized.

leofeyer committed:
Fix the coding style

leofeyer reviewed (changes requested):
Besides the one open question, this PR looks very good. 👍

aschempp commented on core-bundle/src/EventListener/CommandSchedulerListener.php:
```diff
@@ -82,7 +84,8 @@ private function canRunController(Request $request): bool
     private function canRunDbQuery(): bool
     {
         try {
-            return $this->connection->isConnected() && $this->connection->getSchemaManager()->tablesExist(['tl_cron']);
+            return $this->connection->isConnected()
+                && $this->connection->getSchemaManager()->tablesExist(['tl_cron_job']);
```
This sounds like something the `Cron` should check? Though it also sounds like a performance penalty…

fritzmg replied:
No, this is strictly something the `CommandSchedulerListener` should check. The cron should only be executed, if Contao has been installed completely. I can dig up the issue why this was added if you want.

aschempp commented on core-bundle/src/Repository/RememberMeRepository.php:
```diff
@@ -31,9 +31,7 @@ public function __construct(ManagerRegistry $registry)
-        if (($connection = $registry->getConnection()) instanceof Connection) {
-            $this->connection = $connection;
-        }
+        $this->connection = $registry->getConnection();
```
How is this related to the PR ?

fritzmg replied:
Woops, it is not. I accidentally changed this in the wrong repository class.

leofeyer replied:
Nope, that was me. Sorry. The change should have been a reminder to fix this in a separate PR. It was not meant to be committed. 🙈

aschempp commented on core-bundle/src/Resources/config/services.yml:
```diff
@@ -181,6 +185,18 @@ services:
+        arguments:
+            - '@contao.framework'
+        tags:
+            - { name: 'terminal42_service_annotation' }
```
This should not be necessary as the class implements the `ServiceAnnotationInterface` and this config file has `autoconfigure` enabled.

aschempp reviewed (changes requested):
Looks very good to me! I have some minor nitpicks as usual 🙈

aschempp reviewed (changes requested):
I forgot one thing. We should not use the `ScopeAwareCronJobInterface`, but simply pass the scope to the cron service. There cannot be any other argument, as no-one else will call the service.

fritzmg commented:
> I forgot one thing. We should not use the `ScopeAwareCronJobInterface`, but simply pass the scope to the cron service. There cannot be any other argument, as no-one else will call the service.

Yeah, I am in favour of that now too.

fritzmg committed:
make scope mandatory and remove scope interface

fritzmg committed:
remove superfluous service tag

fritzmg committed:
change addCronJob parameter to CronJob object

fritzmg committed:
remove persistAndFlush method from CronJobRepository

fritzmg committed:
fix rebase errors

fritzmg commented:
Further changes:

* `$scope` parameter on `Cron::run` is now mandatory.
* `ScopeAwareCronJobInterface` removed and scope is instead passed to the cron job's method.
* `Cron::addCronJob` now expects a `CronJob` object.
* `CronJobRepository::persistAndFlush` removed.

Toflar reviewed (dismissed):
I like the concept now :)

leofeyer commented on core-bundle/src/Repository/CronJobRepository.php:
```diff
@@ -0,0 +1,50 @@
+    {
+        $table = $this->getClassMetadata()->getTableName();
+
+        $this->connection->exec("LOCK TABLES $table WRITE, $table AS t0 WRITE, $table AS t0_ WRITE");
```
In the `RememberMeRepository`, this statement is just `"LOCK TABLES $table WRITE, $table AS t0_ WRITE"`. Why do we need the additional `t0` here? And why not just `"LOCK TABLES $table WRITE"`?

aschempp replied:
That has already be answered, hasn't it? 😂 Doctrine automatically creates aliases for **every** table internally

leofeyer replied:
Yes, but why do we need `t0` here and not in the `RememberMeRepository` class?

fritzmg replied:
Otherwise I get this exception:

```
  [Doctrine\DBAL\Exception\DriverException]                                                                                                                                                           
  An exception occurred while executing 'SELECT t0.id AS id_1, t0.name AS name_2, t0.lastRun AS lastRun_3 FROM tl_cron_job t0 WHERE t0.name = ? LIMIT 1' with params ["Contao\\CoreBundle\\Cron\\Le   
  gacyCron::onMinutely"]:                                                                                                                                                                             
                                                                                                                                                                                                      
  SQLSTATE[HY000]: General error: 1100 Table 't0' was not locked with LOCK TABLES                                                                                                                     
                                                                                                                                                                                                      

Exception trace:
  at \vendor\doctrine\dbal\lib\Doctrine\DBAL\Driver\AbstractMySQLDriver.php:106
 Doctrine\DBAL\Driver\AbstractMySQLDriver->convertException() at \vendor\doctrine\dbal\lib\Doctrine\DBAL\DBALException.php:166
 Doctrine\DBAL\DBALException::wrapException() at \vendor\doctrine\dbal\lib\Doctrine\DBAL\DBALException.php:146
 Doctrine\DBAL\DBALException::driverExceptionDuringQuery() at \vendor\doctrine\dbal\lib\Doctrine\DBAL\Connection.php:912
 Doctrine\DBAL\Connection->executeQuery() at \vendor\doctrine\orm\lib\Doctrine\ORM\Persisters\Entity\BasicEntityPersister.php:718
 Doctrine\ORM\Persisters\Entity\BasicEntityPersister->load() at \vendor\doctrine\orm\lib\Doctrine\ORM\EntityRepository.php:193
 Doctrine\ORM\EntityRepository->findOneBy() at \vendor\doctrine\orm\lib\Doctrine\ORM\EntityRepository.php:311
 Doctrine\ORM\EntityRepository->resolveMagicCall() at \vendor\doctrine\orm\lib\Doctrine\ORM\EntityRepository.php:228
 Doctrine\ORM\EntityRepository->__call() at \vendor\contao\contao\core-bundle\src\Cron\Cron.php:85
 Contao\CoreBundle\Cron\Cron->run() at \vendor\contao\contao\core-bundle\src\Command\CronCommand.php:50
 Contao\CoreBundle\Command\CronCommand->execute() at \vendor\symfony\console\Command\Command.php:255
 Symfony\Component\Console\Command\Command->run() at \vendor\symfony\console\Application.php:1029
 Symfony\Component\Console\Application->doRunCommand() at \vendor\symfony\framework-bundle\Console\Application.php:97
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at \vendor\symfony\console\Application.php:272
 Symfony\Component\Console\Application->doRun() at \vendor\symfony\framework-bundle\Console\Application.php:83
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at \vendor\symfony\console\Application.php:148
 Symfony\Component\Console\Application->run() at \vendor\contao\contao\manager-bundle\bin\contao-console:38
```

leofeyer replied:
Strange that it works in the `RememberMeRepository` without `t0`. But whatever. 😄

fritzmg replied:
🤷‍♂ 

It seems hacky to me to use `t0`/`t0_` just like that anyway. I mean we cannot really be certain about what Doctrine will actually use, can we?

leofeyer replied:
No, we cannot. 🤷‍♂ Does Doctrine not provide a method to lock tables?

m-vo replied:
Why do you need locking on a table level? Is this only used to guarantee non-concurrent execution of the script?

fritzmg replied:
Yes, otherwise a cron job might be executed twice within the same minute for example.

aschempp commented on core-bundle/src/Resources/config/services.yml:
```diff
@@ -106,7 +106,11 @@ services:
-    Contao\CoreBundle\Controller\FrontendController: ~
+    Contao\CoreBundle\Controller\FrontendController:
+        arguments:
+            - '@Contao\CoreBundle\Cron\Cron'
```
I think this argument is wrong? The cron service is injected on the action, not the constructor.

fritzmg replied:
Fixed in 80a2dab86

fritzmg committed:
remove service argument for FrontendController

leofeyer commented:
Thanks a lot @fritzmg.
